pkg/custompluginmonitor/plugin/plugin.go

```diff
@@ -146,25 +146,52 @@ func (p *Plugin) run(rule cpmtypes.CustomRule) (exitStatus cpmtypes.Status, outp
 	} else {
 		ctx, cancel = context.WithTimeout(context.Background(), *p.config.PluginGlobalConfig.Timeout)
 	}
-	defer cancel()
 
-	cmd := exec.CommandContext(ctx, rule.Path, rule.Args...)
+	// create a process group
+	sysProcAttr := &syscall.SysProcAttr{
+		Setpgid: true,
+	}
+	cmd := exec.Command(rule.Path, rule.Args...)
+	cmd.SysProcAttr = sysProcAttr
 
 	stdoutPipe, err := cmd.StdoutPipe()
 	if err != nil {
+		defer cancel()
 		glog.Errorf("Error creating stdout pipe for plugin %q: error - %v", rule.Path, err)
 		return cpmtypes.Unknown, "Error creating stdout pipe for plugin. Please check the error log"
 	}
 	stderrPipe, err := cmd.StderrPipe()
 	if err != nil {
+		defer cancel()
 		glog.Errorf("Error creating stderr pipe for plugin %q: error - %v", rule.Path, err)
 		return cpmtypes.Unknown, "Error creating stderr pipe for plugin. Please check the error log"
 	}
 	if err := cmd.Start(); err != nil {
+		defer cancel()
 		glog.Errorf("Error in starting plugin %q: error - %v", rule.Path, err)
 		return cpmtypes.Unknown, "Error in starting plugin. Please check the error log"
 	}
 
+	waitChan := make(chan struct{})
+	defer close(waitChan)
+
+	go func() {
+		select {
+		case <-ctx.Done():
+			glog.Errorf("Error in running plugin timeout %q", rule.Path)
+			if cmd.Process == nil || cmd.Process.Pid == 0 {
+				glog.Errorf("Error in cmd.Process check %q", rule.Path)
+				break
+			}
+			err := syscall.Kill(-cmd.Process.Pid, syscall.SIGKILL)
+			if err != nil {
+				glog.Errorf("Error in kill process %d, %v", cmd.Process.Pid, err)
+			}
+		case <-waitChan:
+			cancel()
```

Comment:
How about we just return here? In that way, the cancel code can be unchanged?

Reply:
ok